addons/stock/models/procurement.py

```diff
@@ -35,34 +35,98 @@ class ProcurementRule(models.Model):
-    warehouse_id = fields.Many2one('stock.warehouse', 'Served Warehouse', help='The warehouse this rule is for')
+        help="When ticked, if the move is splitted or cancelled, the next move will be too.")
+    warehouse_id = fields.Many2one('stock.warehouse', 'Warehouse')
     propagate_warehouse_id = fields.Many2one(
         'stock.warehouse', 'Warehouse to Propagate',
         help="The warehouse to propagate on the created move/procurement, which can be different of the warehouse this rule is for (e.g for resupplying rules from another warehouse)")
-
-    def _run_move(self, product_id, product_qty, product_uom, location_id, name, origin, values):
+    auto = fields.Selection([
+        ('manual', 'Manual Operation'),
+        ('transparent', 'Automatic No Step Added')], string='Automatic Move',
+        default='manual', index=True, required=True,
+        help="The 'Manual Operation' value will create a stock move after the current one. "
+             "With 'Automatic No Step Added', the location is replaced in the original move.")
+    show_apply_on = fields.Boolean(compute='_compute_show_apply_on')
+    rule_message = fields.Html(compute='_compute_action_message')
+
+    @api.depends('action')
+    def _compute_show_apply_on(self):
+        for rule in self:
+            if (rule.user_has_groups('base.group_no_one') and rule.action != 'push') or rule.user_has_groups('base.group_multi_company'):
+                rule.show_apply_on = True
+
+    def _get_message_dict(self):
+        if self.action in ('push', 'pull', 'pull_push'):
+            message_dict = {
+                ('pull', True, True): ('When products are needed in %s, %s are created from %s to fulfill the need.') % (self.location_id.display_name, self.picking_type_id.name, self.location_src_id.display_name),
+                ('push', True, True): ('When products arrive at %s, %s are created to send them in %s.') % (self.location_src_id.display_name, self.picking_type_id.name, self.location_id.display_name)
+            }
+        return message_dict
+
+    @api.depends('action', 'location_id', 'location_src_id', 'picking_type_id')
+    def _compute_action_message(self):
+        for rule in self.filtered(lambda rule: rule.location_id and rule.action and rule.picking_type_id):
+            message_dict = rule._get_message_dict()
+            message = message_dict.get((rule.action, bool(rule.location_src_id), bool(rule.location_id))) and message_dict[(rule.action, bool(rule.location_src_id), bool(rule.location_id))] or ""
+            if rule.action == 'pull_push' and rule.location_src_id and rule.location_id:
+                message = message_dict[('pull', bool(rule.location_src_id), bool(rule.location_id))] + "<br/><br/>"\
+                    + message_dict[('push', bool(rule.location_src_id), bool(rule.location_id))]
+            rule.rule_message = message
+
+    def _apply(self, move):
```

Comment:
- Should be multi
- Should check the action_type
- move push _apply from stock_move here, with api.model

Reply:
one rule can be applied to it should a single one 
no need to check action as we already filtered rule based on (push, pull_push)

For third point we have to change all mechanism of applying push rule.... ( is it ok )